packages/labs/src/frameworks/react/create-component.ts

```diff
@@ -0,0 +1,185 @@
+    React.PropsWithChildren<React.PropsWithRef<any>>
+  >
+> = new Map();
+
+/**
+ *  Creates a React component from a CustomElement.
+ */
+// TODO(sorvell): Was unable to remove type E here.
+export const createComponent = <T extends HTMLElement, E>(
+  React: typeof ReactModule,
+  tagName: string,
+  events?: E
+) => {
+  const Component = React.Component;
+  const createElement = React.createElement;
+
+  type ElementProps = Partial<T> & Events<E>;
+
+  type Props = React.PropsWithoutRef<ElementProps> & ForwardedRef;
+
+  // Use cached value if available.
+  let ComponentClass = componentMap.get(tagName);
+  if (ComponentClass !== undefined) {
+    return ComponentClass;
+  }
+
+  class ReactComponent extends Component<Props> {
+    element!: T;
+
+    ref = (element: T) => {
+      this.element = element;
+      // Fulfill any ref we receive in our props.
+      // The `ref` prop is special and cannot be accessed without an error
+      // so use `__forwardedRef` via `React.forwardRef`.
+      const forwardedRef = this.props?.__forwardedRef;
+      if (forwardedRef) {
+        // Note, there are 2 kinds of refs and we manually fulfill them here.
+        // There is no built in React API for this.
+        if (typeof forwardedRef === 'function') {
+          (forwardedRef as (e: T) => void)(element);
+        } else {
+          // This has to be `any` because `current` is readonly.
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (forwardedRef as any).current = element;
+        }
+      }
+    };
+
+    updateElement(oldProps?: Props) {
+      // Set element properties to the values in `this.props`
+      for (const prop in this.props) {
+        setProperty(
+          this.element,
+          prop,
+          this.props[prop as keyof Props],
+          oldProps ? oldProps[prop as keyof Props] : undefined,
+          events
+        );
+      }
+
+      // TODO(sorvell): There's no safe value to set here.
+      // Unset any `oldProps` that are not currently in `this.props`
+      // for (const prop in oldProps) {
+      //   if (!(prop in this.props)) {
+      //     setProperty(this.element, prop, undefined, oldProps[prop], eventNames);
+      //   }
+      // }
+    }
+
+    /**
+     * Updates element properties correctly setting attributes v. properties
+     * on mount.
+     */
+    componentDidMount() {
+      this.updateElement();
+    }
+
+    /**
+     * Updates element properties correctly setting attributes v. properties
+     * on every update. Note, this does not include mount.
+     */
+    componentDidUpdate(old: Props) {
+      this.updateElement(old);
+    }
+
+    /**
+     * Renders creates the custom element with a `ref` prop which allows this
+     * component to reference the custom element.
+     *
+     * Note, element properties are updated in DidMount/DidUpdate.
+     *
+     */
+    render() {
+      return createElement(tagName, {ref: this.ref}, this.props.children);
```

Comment:
why not let the attributes flow through to React's diffing? - just filter out the properties and events?

Reply:
We can't filter out the properties here with a simple `in` check here, since we do not have the element, we'd have to use the element class prototype or lit specific `classProperities`.

Reply:
The thing that's rough with this approach is that in SSR, these elements will have no attributes at all, because `componentDidMount` doesn't run.

We might need to take a config of properties and event names, and then if we want good ergonomics we might want a separate wrapper for when you have the component definition vs when you don't. ie, make a lazy wrapper that takes a config and an eager wrapper that takes the class. The eager wrapper, because it needs to import the definition, will just work. The lazy wrapper will need to take a loader function that performs a dynamic import as an alternative to `React.lazy`.

Reply:
Now computes a list of properties and events that are handled specially and separates the user's `props` into a list that's passed to React (so that React's rules for attribute setting are used) and a list of properties and events that's handled specially by the wrapper.